### PR TITLE
Histogram Aggregation: Fix bucket detection logic, performance improvements, and benchmark tests

### LINKED ISSUE #1864: Why boundaries is std::list
### Discussed in https://github.com/open-telemetry/opentelemetry-cpp/discussions/1863

<div type='discussions-op-text'>

<sup>Originally posted by **ondra-novak** December 14, 2022</sup>
https://github.com/open-telemetry/opentelemetry-cpp/blob/b8b715faf9a364767bb851de763ccef983d65ccf/sdk/include/opentelemetry/sdk/metrics/data/point_data.h#L59

Just explain me, why list is used here. Unless you need to insert into or erase from inside of the list, this container is very slow and you are loosing a lot of performance here.

This has been changed from std::vector in commit a7e814a632fdaee0c76246358ffad811ac06a7e9 without any sign of need such a change and without explanation. 

</div>

### LINKED ISSUE #1865: [Metrics] UInt64Histogram does not honour custom boundaries
If you `CreateUInt64Histogram` and add custom boundaries using a view, the counts output doesn't match the expected output.
**Note** `CreateDoubleHistogram` works as expected 

This unit test can demonstrate the problem and expected output.
``` C++
TEST(MeterProvider, UInt64HistogramBuckets)
{
  MeterProvider mp;
  auto m = mp.GetMeter("meter1", "version1", "schema1");

  std::unique_ptr<MockMetricExporter> exporter(new MockMetricExporter());
  std::shared_ptr<MetricReader> reader{new MockMetricReader(std::move(exporter))};
  mp.AddMetricReader(reader);

  std::shared_ptr<HistogramAggregationConfig> config(new HistogramAggregationConfig());
  config->boundaries_ = {10, 20, 30, 40};
  std::unique_ptr<View> view{new View("view1", "view1_description", AggregationType::kHistogram, config)};
  std::unique_ptr<InstrumentSelector> instrument_selector{
      new InstrumentSelector(InstrumentType::kHistogram, "histogram1")};
  std::unique_ptr<MeterSelector> meter_selector{new MeterSelector("meter1", "version1", "schema1")};
  mp.AddView(std::move(instrument_selector), std::move(meter_selector), std::move(view));

  auto h = m->CreateUInt64Histogram("histogram1", "histogram1_description", "histogram1_unit");

  h->Record(10, {});
  h->Record(20, {});
  h->Record(30, {});
  h->Record(40, {});
  h->Record(50, {});

  std::list<double> actual_boundaries;
  std::vector<uint64_t> actual_counts;
  reader->Collect([&](ResourceMetrics &rm) {
    for(const ScopeMetrics& smd : rm.scope_metric_data_) {
      for(const MetricData& md : smd.metric_data_) {
        for(const PointDataAttributes& dp : md.point_data_attr_) {
          auto data = opentelemetry::nostd::get<HistogramPointData>(dp.point_data);
          actual_boundaries = data.boundaries_;
          actual_counts = data.counts_;
        }
      }
    }
    return true;
  });
  ASSERT_EQ(std::list<double>({10, 20, 30, 40}), actual_boundaries);
  ASSERT_EQ(std::vector<uint64_t>({1, 1, 1, 1, 1}), actual_counts);
}
```

The output of this test is:
```
Expected equality of these values:
  std::vector<uint64_t>({1, 1, 1, 1, 1})
    Which is: { 1, 1, 1, 1, 1 }
  actual_counts
    Which is: { 0, 1, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0 }
```

The `boundaries_` values are correct, but the `counts_` are wrong
16 count values is the expected size for the default set of boundaries, but the counts don't match default boundaries.

This has been tested against release 1.7.0, 1.8.1 and commit e7579ed68869cd3aa7a639c239127bec7d6334ce

### PR DESCRIPTION
**Pull request by lalitb**
Fixes #1864, #1865, #1866

## Changes

1. Fixed the bucket detection logic
    -  Lower bound value is exclusive, Upper bound value is inclusive ( as mentioned in [specs ](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/sdk.md#explicit-bucket-histogram-aggregation))
    - The last bucket was not getting populated.
3. The default boundaries for UINT64 was incorrect, changed it to use the values as specified in above specs link
4. Added separate tests for histogram buckets with default and custom boundaries (shamelessly taken from the issue).
5. Pass custom boundaries config data while merging and diff-ing Histogram aggregation.
6. Change boundaries data type from std::list to std::vector. And use binary search for searching the right bucket for a measurement.
7. Add histogram performance benchmark tests.

Benchmark result for histogram aggregation (prior to changes in point 5 above):
```
$ ./histogram_aggregation_benchmark
2022-12-16 15:18:01
Running ./histogram_aggregation_benchmark
Run on (20 X 2808.01 MHz CPU s)
CPU Caches:
  L1 Data 32K (x10)
  L1 Instruction 32K (x10)
  L2 Unified 256K (x10)
  L3 Unified 20480K (x1)
Load Average: 2.90, 1.64, 0.61
------------------------------------------------------------------
Benchmark                        Time             CPU   Iterations
------------------------------------------------------------------
BM_HistogramAggregation        740 ns          740 ns       936807
```
Benchmark result for histogram aggregation (after changes in point 5 above):
```
$ ./histogram_aggregation_benchmark
2022-12-16 15:17:58
Running ./histogram_aggregation_benchmark
Run on (20 X 2808.01 MHz CPU s)
CPU Caches:
  L1 Data 32K (x10)
  L1 Instruction 32K (x10)
  L2 Unified 256K (x10)
  L3 Unified 20480K (x1)
Load Average: 2.90, 1.64, 0.61
------------------------------------------------------------------
Benchmark                        Time             CPU   Iterations
------------------------------------------------------------------
BM_HistogramAggregation        552 ns          552 ns      1297628
```

For significant contributions please make sure you have completed the following items:

* [ ] `CHANGELOG.md` updated for non-trivial changes
* [x] Unit tests have been added
* [x] Changes in public API reviewed

### CODE REVIEW

File: sdk/test/metrics/histogram_test.cc [hunk 1]
```diff
@@ -0,0 +1,278 @@
...
+
+  ASSERT_EQ(1, actuals.size());
+
+  const auto &actual = actuals.at(0);
+  ASSERT_EQ(275, opentelemetry::nostd::get<int64_t>(actual.sum_));
+  ASSERT_EQ(10, actual.count_);
+  ASSERT_EQ(5, opentelemetry::nostd::get<int64_t>(actual.min_));
+  ASSERT_EQ(50, opentelemetry::nostd::get<int64_t>(actual.max_));
+  ASSERT_EQ(std::list<double>({10, 20, 30, 40}), actual.boundaries_);
+  ASSERT_EQ(std::vector<uint64_t>({2, 2, 2, 2, 2, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0}),
```
Comment (sam-leitch-oxb) [inline sdk/test/metrics/histogram_test.cc:276]: Shouldn't this be `{2, 2, 2, 2, 2}`?

Comment (lalitb) [reply sdk/test/metrics/histogram_test.cc]: Indeed it should be. Have fixed the code for that. Thanks for pointing out.

File: exporters/otlp/test/otlp_metrics_serialization_test.cc [hunk 2]
```diff
@@ -5,6 +5,7 @@
 #include "opentelemetry/proto/metrics/v1/metrics.pb.h"
 
 #include <gtest/gtest.h>
+#include <vector>
```
Comment (ThomsonTan) [inline exporters/otlp/test/otlp_metrics_serialization_test.cc:8]: Why `std::list` works but not `std::vector`?

Comment (lalitb) [reply exporters/otlp/test/otlp_metrics_serialization_test.cc]: Oh I don't remember, have remove the header now :)

File: sdk/include/opentelemetry/sdk/metrics/state/temporal_metric_storage.h [hunk 3]
```diff
@@ -8,6 +8,7 @@
 #include "opentelemetry/sdk/metrics/state/attributes_hashmap.h"
 #include "opentelemetry/sdk/metrics/state/metric_collector.h"
 
+#include <list>
```
Comment (ThomsonTan) [inline sdk/include/opentelemetry/sdk/metrics/state/temporal_metric_storage.h:11]: Is this needed?

File: sdk/include/opentelemetry/sdk/metrics/aggregation/histogram_aggregation.h [hunk 4]
```diff
@@ -108,6 +108,26 @@ void HistogramDiff(HistogramPointData &current, HistogramPointData &next, Histog
   diff.record_min_max_ = false;
 }
 
+template <class T>
+size_t BucketBinarySearch(T value, std::vector<double> &boundaries)
```
Comment (ThomsonTan) [inline sdk/include/opentelemetry/sdk/metrics/aggregation/histogram_aggregation.h:112]: Can we use std::binary_search here?

Comment (lalitb) [reply sdk/include/opentelemetry/sdk/metrics/aggregation/histogram_aggregation.h:112]: “std::binary_search” will not provide index of element for insertion, it just provides true or false if element is found.

File: sdk/include/opentelemetry/sdk/metrics/state/temporal_metric_storage.h (hunk 3 above)
Comment (lalitb) [reply sdk/include/opentelemetry/sdk/metrics/state/temporal_metric_storage.h:11]: yes, it is used to maintain the list of unreported metrics :)

**Review by ThomsonTan — APPROVED** [review]

**Comment by lalitb** [comment]
Can I merge this PR, as this has few trivial bug fixes, and performance improvement for histogram aggregation, don't want to keep this on hold for long. Though it is approved, I realize it is still not meeting our policy, so checking if there are any issue :) I can incorporate any later comments as separate PR if required.

``
A pull request opened by an Approver / Maintainer can be merged with only one approval from Approver / Maintainer (at different company).
``

p.s. - As special case, we have been merging ETW exporter related PRs and few smaller PRs with one approval from same company, but that should be acceptable.

**Comment by ThomsonTan** [comment]
> Can I merge this PR, as this has few trivial bug fixes, and performance improvement for histogram aggregation, don't want to keep this on hold for long. Though it is approved, I realize it is still not meeting our merge policy, so checking if there are any issue :) I can incorporate any later comments as separate PR if required.
> 
> `A pull request opened by an Approver / Maintainer can be merged with only one approval from Approver / Maintainer (at different company).`
> 
> p.s. - As special case, we have been merging ETW exporter related PRs and few smaller PRs with one approval from same company, but that should be acceptable.

Thanks for raising this. @esigo could you please help take a look when you have time?

For ETW exporter, it seems more appropriate to move it to contrib repo.

File: sdk/include/opentelemetry/sdk/metrics/aggregation/histogram_aggregation.h [hunk 5]
```diff
@@ -108,6 +108,26 @@ void HistogramDiff(HistogramPointData &current, HistogramPointData &next, Histog
...
+    {
+      lower_bound = mid + 1;
+    }
+    else
+    {
+      upper_bound = mid;
+    }
+  }
+  return lower_bound;
+}
```
Comment (esigo) [inline sdk/include/opentelemetry/sdk/metrics/aggregation/histogram_aggregation.h:129]:
```suggestion
template <class T>
size_t BucketBinarySearch(T value, const std::vector<double> &boundaries)
{
    auto low = std::lower_bound(boundaries.begin(), boundaries.end(), value);
    return low - boundaries.begin();
}
```

**Review by esigo — APPROVED** [review]
LGTM
Thanks for the PR. Just had a suggestion :)
Sorry for the delay.

File: sdk/include/opentelemetry/sdk/metrics/aggregation/histogram_aggregation.h (hunk 5 above)
Comment (lalitb) [reply sdk/include/opentelemetry/sdk/metrics/aggregation/histogram_aggregation.h]: Nice change. Thanks.

**Comment by lalitb** [comment]
> LGTM Thanks for the PR. Just had a suggestion :) Sorry for the delay.

No issues. Thanks for reviewing :)

**Pull request merged**